lib/propolis/src/hw/nvme/mod.rs

```diff
@@ -760,17 +760,41 @@ impl PciNvme {
             }
 
             CtrlrReg::DoorBellAdminSQ => {
-                let val = wo.read_u32().try_into().unwrap();
+                // 32-bit register but ignore reserved top 16-bits
+                let val = wo.read_u32() as u16;
                 let state = self.state.lock().unwrap();
+
+                if !state.ctrl.cc.enabled() {
+                    slog::warn!(
+                        self.log,
+                        "Doorbell write while controller is disabled"
+                    );
+                    return Err(NvmeError::InvalidSubQueue(
+                        queue::ADMIN_QUEUE_ID,
+                    ));
+                }
+
                 let admin_sq = state.get_admin_sq();
```

Comment:
What about making `get_admin_sq` and `get_admin_cq` return an `Option<>` instead of checking the state bits manually?

Reply:
I'd like to keep the controller-enabled check cause those are slightly different conditions.

But good point on `get_admin_{cq,sq}`. In theory we shouldn't encounter a doorbell reg write while the controller is active but the admin queues haven't been created. But we shouldn't panic either. I switched them to just return a `Result` like the `get_{cq,sq}` they're wrapping.